node/impl/full/actor_events.go

```diff
@@ -0,0 +1,226 @@
+package full
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	"go.uber.org/fx"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/events/filter"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type ActorEventAPI interface {
+	GetActorEvents(ctx context.Context, filter *types.ActorEventFilter) ([]*types.ActorEvent, error)
+	SubscribeActorEvents(ctx context.Context, filter *types.SubActorEventFilter) (<-chan *types.ActorEvent, error)
+}
+
+var (
+	_ ActorEventAPI = *new(api.FullNode)
+	_ ActorEventAPI = *new(api.Gateway)
+)
+
+type ActorEventHandler struct {
+	EventFilterManager   *filter.EventFilterManager
+	MaxFilterHeightRange abi.ChainEpoch
+	Chain                *store.ChainStore
+}
+
+var _ ActorEventAPI = (*ActorEventHandler)(nil)
+
+type ActorEventsAPI struct {
+	fx.In
+	ActorEventAPI
+}
+
+func (a *ActorEventHandler) GetActorEvents(ctx context.Context, filter *types.ActorEventFilter) ([]*types.ActorEvent, error) {
```

Comment:
We should have a limit here, although we may want to install that limit in the gateway? Do we even expose this API on the gateway?

I.e., I'm thinking:

1. A max number of epochs.
2. A max number of events to collect before aborting.

But, given those limits... we may want to change the API? E.g.:

1. Return an object specifying the range of epochs we returned events from.
2. Guarantee that we'll either return all events from an epoch, or none.

(I'm starting to become more convinced that we should have batched by epoch...).

Reply:
Really, we likely need:

1. A `Limit` field in the `ActorEventFilter` (max number of events).
2. Something in the _gateway_ that caps the `Limit`.

Reply:
Already have the latter, `cfg.Fevm.Events.MaxFilterHeightRange` exists already as part of fevm events and we're piggy backing that, defaults to 2880. Because it's old, it's going to be difficult to move it to the new top-level `Events` object but it would be nice for them to sit together ... maybe we need to consider a config "migration", or at least having a config unifying thing run at startup to support old + new for these events values.

Reply:
That's probably fine? But then we need to make sure to apply that in the gateway proxy for the events API.

Or are we applying that deep in the system? If so... ugh, that's not how we do things... but oh well.

Reply:
No, thankfully not, it's at the parse-inputs phase, which is one main purpose of the `parseBlockRange`, which itself was an extraction of existing code for fevm event filter code.

What I'm doing now is keeping that function intact for the fevm case so we leave that functionality as is, but implementing a simpler one for `parseHeightRange` for this API that applies the same limit but also gets us from inputs to the internal filter representation. That turns out to be a bit tricky to unify because the `"earliest"` and `"latest"` stuff comes from the existing fevm events filter API. So I'm just trying to make sure tests cover the edges of the conversions.